examples/step-74/step-laplace-FEIV.cc

```diff
@@ -0,0 +1,1159 @@
+    }
+  };
+
+  template <class MatrixType, class VectorType>
+  inline void copy(const CopyData &                 c,
+                   const AffineConstraints<double> &constraints,
+                   MatrixType &                     system_matrix,
+                   VectorType &                     system_rhs)
+  {
+    constraints.distribute_local_to_global(c.cell_matrix,
+                                           c.cell_rhs,
+                                           c.local_dof_indices,
+                                           system_matrix,
+                                           system_rhs);
+    for (auto &cdf : c.face_data)
+      {
+        const unsigned int dofs_per_cell = cdf.joint_dof_indices.size();
+        for (unsigned int i = 0; i < dofs_per_cell; ++i)
+          for (unsigned int k = 0; k < dofs_per_cell; ++k)
+            system_matrix.add(cdf.joint_dof_indices[i],
+                              cdf.joint_dof_indices[k],
+                              cdf.cell_matrix(i, k));
+      }
+  }
+
+
+  template <int dim>
+  class SIPGLaplace
+  {
+  public:
+    SIPGLaplace(const Test_Case &test_case);
+    void run();
+
+  private:
+    void   setup_system();
+    void   assemble_system();
+    void   solve();
+    void   refine_grid();
+    void   output_results(const unsigned int cycle) const;
+    double compute_penalty(const double h1, const double h2);
+    void   compute_errors();
+    void   compute_error_estimate();
+    double compute_energy_norm();
+    // The following functions will be moved to FEIV
+    void get_function_values(const FEInterfaceValues<dim> &fe_iv,
+                             const Vector<double> &        solution,
+                             std::vector<double>           face_values[2]);
+    void get_function_jump(const FEInterfaceValues<dim> &fe_iv,
+                           const Vector<double> &        solution,
+                           std::vector<double> &         jump);
+    void get_function_average(const FEInterfaceValues<dim> &fe_iv,
+                              const Vector<double> &        solution,
+                              std::vector<double> &         average);
+    void get_function_gradients(const FEInterfaceValues<dim> &fe_iv,
+                                const Vector<double> &        solution,
+                                std::vector<Tensor<1, dim>> face_gradients[2]);
+    void get_function_gradient_jump(const FEInterfaceValues<dim> &fe_iv,
+                                    const Vector<double> &        solution,
+                                    std::vector<Tensor<1, dim>> &gradient_jump);
+    void get_function_gradient_average(
+      const FEInterfaceValues<dim> &fe_iv,
+      const Vector<double> &        solution,
+      std::vector<Tensor<1, dim>> & gradient_average);
+    const Test_Case      test_case;
+    Triangulation<dim>   triangulation;
+    const MappingQ1<dim> mapping;
+
+    AffineConstraints<double> constraints;
```

Comment:
Maybe this is a stupid comment, but I guess you do not impose Dirichlets BCs strongly. Then, why is there a need for constraints?

Reply:
It is only used to distribute local to global: constraints.distribute_local_to_global(...)